File: bench/src/collatz.pl
:- use_module(library(tabling)).

collatz(A, B) :-
	  0 is A mod 2, B is A // 2
	; 1 is A mod 2, B is 3 * A + 1.

:- table collatz_length/2.
collatz_length(1, 0) :- !.
collatz_length(N0, I0) :-
	collatz(N0, N),
	collatz_length(N, I),
	I0 is I + 1.

max_collatz_state(s(I0, Best0, Len0), s(I, Best, Len)) :-
	I is I0 + 1,
	collatz_length(I, L) ->
	  ( L >  Len0, !, Len = L, Best = I
	  ; L =< Len0, !, Len = Len0, Best = Best0 ).

reachable(_, S, S).
reachable(F, S0, S) :-
	call(F, S0, S1),
	reachable(F, S1, S).

max_collatz(Lim, N, M) :-
	reachable(max_collatz_state, s(0, 0, 0), s(Lim, N, M)).

:- initialization((max_collatz(1000000, _, M) -> write(M), halt)).

File: bench/src/collatz.ml
let collatz n =
  let rec f n i =
    match n with
    | n when n <= 1 -> i
    | n when n mod 2 = 0 -> f  (n / 2)  (i + 1)
    | n when n mod 2 = 1 -> f (3*n + 1) (i + 1)
    | _ -> 0
  in f n 0
;;

let maxlen lim =
  let rec f lim best =
    if lim <= 1 then best
    else let b = collatz lim in
      f (lim - 1) (if b > best then b else best)
  in f lim 0
;;

Printf.printf "%d" (maxlen 1000000)
